File: Windows/src/IwashiScope.Core/Calculations/MunsellConverter.cs
using System.Globalization;
using System.Reflection;
using IwashiScope.Core.Models;

namespace IwashiScope.Core.Calculations;

public sealed record MunsellNotation(
    double? Hue,
    string? HueDesignator,
    double Value,
    double Chroma)
{
    public string Formatted => Hue is { } hue && HueDesignator is { } designator
        ? $"{Number(hue)}{designator} {Number(Value)}/{Number(Chroma)}"
        : $"N {Number(Value)}";

    private static string Number(double value)
    {
        var rounded = Math.Round(value, 1, MidpointRounding.AwayFromZero);
        return Math.Abs(rounded - Math.Round(rounded)) < 0.000_001
            ? Math.Round(rounded).ToString("0", CultureInfo.InvariantCulture)
            : rounded.ToString("0.0", CultureInfo.InvariantCulture);
    }
}

public static class MunsellConverter
{
    private readonly record struct Tristimulus(double X, double Y, double Z);
    private readonly record struct ColorimetrySample(
        double Wavelength,
        double Illuminant,
        double XBar,
        double YBar,
        double ZBar);
    private readonly record struct ReflectancePoint(double Wavelength, double Value);
    private readonly record struct IntegrationPoint(
        double Wavelength,
        double X,
        double Y,
        double Z,
        double Normalization);
    private readonly record struct LabPoint(double L, double A, double B)
    {
        public double Chroma => Math.Sqrt(A * A + B * B);
        public bool IsFinite => double.IsFinite(L) && double.IsFinite(A) && double.IsFinite(B);
    }
    private sealed record RenotationSample(
        double AstmHue,
        double Value,
        double Chroma,
        LabPoint Lab);

    private static readonly Tristimulus IlluminantCWhite =
        new(0.9807059717, 1, 1.1822494939);
    private static readonly Lazy<IReadOnlyList<RenotationSample>> RenotationSamples =
        new(LoadRenotationSamples);
    private static readonly Lazy<IReadOnlyList<ColorimetrySample>> ColorimetrySamples =
        new(LoadColorimetrySamples);
    private const int NeighborCount = 12;
    private const double NeutralChromaThreshold = 0.5;

    public static MunsellNotation? Convert(IReadOnlyList<SpectralSample> reflectanceSpectrum)
    {
        var xyz = IlluminantCXYZ(reflectanceSpectrum);
        if (xyz is null || RenotationSamples.Value.Count == 0)
        {
            return null;
        }

        var normalizedXyz = new Tristimulus(
            xyz.First / 100,
            xyz.Second / 100,
            xyz.Third / 100);
        var target = LabFromXyz(normalizedXyz, IlluminantCWhite);
        if (!target.IsFinite || !xyz.IsFinite)
        {
            return null;
        }

        var value = MunsellValue(xyz.Second);
        var nearbyByValue = RenotationSamples.Value
            .Where(sample => Math.Abs(sample.Value - value) <= 1.1)
            .ToArray();
        var pool = nearbyByValue.Length > 0 ? nearbyByValue : RenotationSamples.Value;
        var neighbors = pool
            .Select(sample =>
            {
                var deltaL = sample.Lab.L - target.L;
                var deltaA = sample.Lab.A - target.A;
                var deltaB = sample.Lab.B - target.B;
                return (
                    Sample: sample,
                    SquaredDistance: deltaL * deltaL + deltaA * deltaA + deltaB * deltaB);
            })
            .OrderBy(candidate => candidate.SquaredDistance)
            .Take(NeighborCount)
            .ToArray();
        if (neighbors.Length == 0)
        {
            return null;
        }

        var sine = 0d;
        var cosine = 0d;
        var chromaScale = 0d;
        var totalWeight = 0d;
        foreach (var neighbor in neighbors)
        {
            var weight = 1 / (neighbor.SquaredDistance + 0.000_001);
            var angle = neighbor.Sample.AstmHue * 2 * Math.PI / 100;
            sine += weight * Math.Sin(angle);
            cosine += weight * Math.Cos(angle);
            chromaScale += weight * neighbor.Sample.Chroma /
                Math.Max(neighbor.Sample.Lab.Chroma, 0.000_001);
            totalWeight += weight;
        }
        if (!double.IsFinite(totalWeight) || totalWeight <= 0)
        {
            return null;
        }

        var chroma = Math.Clamp(target.Chroma * chromaScale / totalWeight, 0, 50);
        if (chroma < NeutralChromaThreshold)
        {
            return new MunsellNotation(null, null, value, 0);
        }

        var astmHue = Math.Atan2(sine, cosine) * 100 / (2 * Math.PI);
        if (astmHue <= 0)
        {
            astmHue += 100;
        }
        var (prefix, designator) = HueNotation(astmHue);
        return new MunsellNotation(prefix, designator, value, chroma);
    }

    public static Vector3? IlluminantCXYZ(IReadOnlyList<SpectralSample> reflectanceSpectrum)
    {
        var reflectancePoints = NormalizedReflectancePoints(reflectanceSpectrum);
        var references = ColorimetrySamples.Value;
        if (reflectancePoints.Count < 2 || references.Count == 0)
        {
            return null;
        }

        var lowerBound = Math.Max(
            reflectancePoints[0].Wavelength,
            references[0].Wavelength);
        var upperBound = Math.Min(
            reflectancePoints[^1].Wavelength,
            references[^1].Wavelength);
        if (upperBound - lowerBound < 1)
        {
            return null;
        }

        var wavelengths = new List<double> { lowerBound };
        wavelengths.AddRange(references
            .Select(sample => sample.Wavelength)
            .Where(wavelength => wavelength > lowerBound && wavelength < upperBound));
        wavelengths.Add(upperBound);

        var xIntegral = 0d;
        var yIntegral = 0d;
        var zIntegral = 0d;
        var normalizationIntegral = 0d;
        IntegrationPoint? previous = null;
        foreach (var wavelength in wavelengths)
        {
            var reflectance = InterpolateReflectance(reflectancePoints, wavelength);
            var reference = InterpolateColorimetry(wavelength);
            if (reflectance is null || reference is null)
            {
                return null;
            }

            var scaledReflectance = Math.Max(reflectance.Value / 100, 0);
            var current = new IntegrationPoint(
                wavelength,
                scaledReflectance * reference.Value.Illuminant * reference.Value.XBar,
                scaledReflectance * reference.Value.Illuminant * reference.Value.YBar,
                scaledReflectance * reference.Value.Illuminant * reference.Value.ZBar,
                reference.Value.Illuminant * reference.Value.YBar);
            if (previous is { } prior)
            {
                var interval = current.Wavelength - prior.Wavelength;
                xIntegral += interval * (prior.X + current.X) / 2;
                yIntegral += interval * (prior.Y + current.Y) / 2;
                zIntegral += interval * (prior.Z + current.Z) / 2;
                normalizationIntegral += interval
                    * (prior.Normalization + current.Normalization) / 2;
            }
            previous = current;
        }

        if (!double.IsFinite(normalizationIntegral) || normalizationIntegral <= 0)
        {
            return null;
        }
        var scale = 100 / normalizationIntegral;
        var result = new Vector3(
            xIntegral * scale,
            yIntegral * scale,
            zIntegral * scale);
        return result.IsFinite ? result : null;
    }

    private static IReadOnlyList<ReflectancePoint> NormalizedReflectancePoints(
        IReadOnlyList<SpectralSample> spectrum)
    {
        var sorted = spectrum
            .Where(sample => double.IsFinite(sample.Wavelength) && double.IsFinite(sample.Value))
            .OrderBy(sample => sample.Wavelength);
        var points = new List<ReflectancePoint>();
        foreach (var sample in sorted)
        {
            var point = new ReflectancePoint(sample.Wavelength, sample.Value);
            if (points.Count > 0 &&
                Math.Abs(points[^1].Wavelength - point.Wavelength) < 0.000_000_001)
            {
                points[^1] = point;
            }
            else
            {
                points.Add(point);
            }
        }
        return points;
    }

    private static double? InterpolateReflectance(
        IReadOnlyList<ReflectancePoint> points,
        double wavelength)
    {
        if (points.Count == 0 ||
            wavelength < points[0].Wavelength ||
            wavelength > points[^1].Wavelength)
        {
            return null;
        }
        var lower = 0;
        var upper = points.Count - 1;
        while (lower + 1 < upper)
        {
            var midpoint = (lower + upper) / 2;
            if (points[midpoint].Wavelength <= wavelength)
            {
                lower = midpoint;
            }
            else
            {
                upper = midpoint;
            }
        }
        var first = points[lower];
        if (Math.Abs(first.Wavelength - wavelength) < 0.000_000_001)
        {
            return first.Value;
        }
        var second = points[upper];
        var interval = second.Wavelength - first.Wavelength;
        if (interval <= 0)
        {
            return first.Value;
        }
        var fraction = (wavelength - first.Wavelength) / interval;
        return first.Value + fraction * (second.Value - first.Value);
    }

    private static ColorimetrySample? InterpolateColorimetry(double wavelength)
    {
        var samples = ColorimetrySamples.Value;
        if (samples.Count == 0 ||
            wavelength < samples[0].Wavelength ||
            wavelength > samples[^1].Wavelength)
        {
            return null;
        }
        var position = wavelength - samples[0].Wavelength;
        var lowerIndex = Math.Clamp((int)Math.Floor(position), 0, samples.Count - 1);
        var lower = samples[lowerIndex];
        if (lowerIndex == samples.Count - 1 ||
            Math.Abs(lower.Wavelength - wavelength) < 0.000_000_001)
        {
            return lower;
        }
        var upper = samples[lowerIndex + 1];
        var fraction = (wavelength - lower.Wavelength)
            / (upper.Wavelength - lower.Wavelength);
        return new ColorimetrySample(
            wavelength,
            lower.Illuminant + fraction * (upper.Illuminant - lower.Illuminant),
            lower.XBar + fraction * (upper.XBar - lower.XBar),
            lower.YBar + fraction * (upper.YBar - lower.YBar),
            lower.ZBar + fraction * (upper.ZBar - lower.ZBar));
    }

    private static (double Prefix, string Designator) HueNotation(double astmHue)
    {
        string[] designators = ["R", "YR", "Y", "GY", "G", "BG", "B", "PB", "P", "RP"];
        var normalized = astmHue % 100;
        if (normalized <= 0) normalized += 100;
        var segment = (int)Math.Floor(normalized / 10);
        var prefix = normalized - segment * 10;
        if (prefix < 0.000_001)
        {
            segment = (segment + designators.Length - 1) % designators.Length;
            prefix = 10;
        }
        else
        {
            segment %= designators.Length;
        }
        return (prefix, designators[segment]);
    }

    private static double MunsellValue(double luminance)
    {
        var target = Math.Clamp(luminance, 0, 100);
        var lower = 0d;
        var upper = 10d;
        for (var iteration = 0; iteration < 60; iteration++)
        {
            var candidate = (lower + upper) / 2;
            if (AstmLuminance(candidate) < target)
            {
                lower = candidate;
            }
            else
            {
                upper = candidate;
            }
        }
        return (lower + upper) / 2;
    }

    private static double AstmLuminance(double value) =>
        0.00081939 * Math.Pow(value, 5) -
        0.020484 * Math.Pow(value, 4) +
        0.23352 * Math.Pow(value, 3) -
        0.22533 * Math.Pow(value, 2) +
        1.1914 * value;

    private static LabPoint LabFromXyz(Tristimulus xyz, Tristimulus white)
    {
        var fx = LabPivot(xyz.X / white.X);
        var fy = LabPivot(xyz.Y / white.Y);
        var fz = LabPivot(xyz.Z / white.Z);
        return new LabPoint(116 * fy - 16, 500 * (fx - fy), 200 * (fy - fz));
    }

    private static double LabPivot(double value)
    {
        const double epsilon = 216.0 / 24389.0;
        const double kappa = 24389.0 / 27.0;
        return value > epsilon ? Math.Cbrt(value) : (kappa * value + 16) / 116;
    }

    private static IReadOnlyList<RenotationSample> LoadRenotationSamples()
    {
        using var stream = Assembly.GetExecutingAssembly()
            .GetManifestResourceStream("IwashiScope.MunsellRenotationAll.csv");
        if (stream is null)
        {
            return [];
        }
        using var reader = new StreamReader(stream);
        _ = reader.ReadLine();
        var samples = new List<RenotationSample>(4_995);
        while (reader.ReadLine() is { } line)
        {
            var fields = line.Split(',');
            if (fields.Length != 6 ||
                !double.TryParse(fields[1], NumberStyles.Float, CultureInfo.InvariantCulture, out var value) ||
                !double.TryParse(fields[2], NumberStyles.Float, CultureInfo.InvariantCulture, out var chroma) ||
                !double.TryParse(fields[3], NumberStyles.Float, CultureInfo.InvariantCulture, out var x) ||
                !double.TryParse(fields[4], NumberStyles.Float, CultureInfo.InvariantCulture, out var y) ||
                !double.TryParse(fields[5], NumberStyles.Float, CultureInfo.InvariantCulture, out var rawY) ||
                Math.Abs(y) <= 0.000_000_001 ||
                AstmHue(fields[0]) is not { } astmHue)
            {
                continue;
            }

            var luminance = rawY * 0.975 / 100;
            var xyz = new Tristimulus(
                x * luminance / y,
                luminance,
                (1 - x - y) * luminance / y);
            var lab = LabFromXyz(xyz, IlluminantCWhite);
            if (!lab.IsFinite || lab.Chroma <= 0.000_001)
            {
                continue;
            }
            samples.Add(new RenotationSample(astmHue, value, chroma, lab));
        }
        return samples;
    }

    private static double? AstmHue(string label)
    {
        string[] designators = ["BG", "GY", "YR", "RP", "PB", "B", "G", "Y", "R", "P"];
        var codes = new Dictionary<string, int>
        {
            ["BG"] = 2, ["GY"] = 4, ["YR"] = 6, ["RP"] = 8, ["PB"] = 10,
            ["B"] = 1, ["G"] = 3, ["Y"] = 5, ["R"] = 7, ["P"] = 9,
        };
        var designator = designators.FirstOrDefault(candidate => label.EndsWith(candidate, StringComparison.Ordinal));
        if (designator is null ||
            !double.TryParse(
                label[..^designator.Length],
                NumberStyles.Float,
                CultureInfo.InvariantCulture,
                out var prefix))
        {
            return null;
        }
        var hue = 10 * ((7 - codes[designator] + 10) % 10) + prefix;
        return Math.Abs(hue) < 0.000_001 ? 100 : hue;
    }

    private static IReadOnlyList<ColorimetrySample> LoadColorimetrySamples()
    {
        using var stream = Assembly.GetExecutingAssembly()
            .GetManifestResourceStream("IwashiScope.MunsellColorimetryCIE1931.csv");
        if (stream is null)
        {
            return [];
        }
        using var reader = new StreamReader(stream);
        _ = reader.ReadLine();
        var samples = new List<ColorimetrySample>(421);
        while (reader.ReadLine() is { } line)
        {
            var fields = line.Split(',');
            if (fields.Length != 5 ||
                !double.TryParse(fields[0], NumberStyles.Float, CultureInfo.InvariantCulture, out var wavelength) ||
                !double.TryParse(fields[1], NumberStyles.Float, CultureInfo.InvariantCulture, out var illuminant) ||
                !double.TryParse(fields[2], NumberStyles.Float, CultureInfo.InvariantCulture, out var xBar) ||
                !double.TryParse(fields[3], NumberStyles.Float, CultureInfo.InvariantCulture, out var yBar) ||
                !double.TryParse(fields[4], NumberStyles.Float, CultureInfo.InvariantCulture, out var zBar))
            {
                continue;
            }
            samples.Add(new ColorimetrySample(
                wavelength,
                illuminant,
                xBar,
                yBar,
                zBar));
        }
        return samples.OrderBy(sample => sample.Wavelength).ToArray();
    }
}

File: IwashiScopePackage/Sources/IwashiScopeFeature/Models/MunsellConverter.swift
import Foundation

struct MunsellNotation: Equatable, Sendable {
    let hue: Double?
    let hueDesignator: String?
    let value: Double
    let chroma: Double

    var formatted: String {
        guard let hue, let hueDesignator else {
            return "N \(Self.number(value))"
        }
        return "\(Self.number(hue))\(hueDesignator) \(Self.number(value))/\(Self.number(chroma))"
    }

    private static func number(_ value: Double) -> String {
        let rounded = (value * 10).rounded() / 10
        if abs(rounded - rounded.rounded()) < 0.000_001 {
            return String(Int(rounded.rounded()))
        }
        return String(format: "%.1f", locale: Locale(identifier: "en_US_POSIX"), rounded)
    }
}

enum MunsellConverter {
    private struct ColorimetrySample {
        let wavelength: Double
        let illuminant: Double
        let xBar: Double
        let yBar: Double
        let zBar: Double
    }

    private struct ReflectancePoint {
        let wavelength: Double
        let value: Double
    }

    private struct LabPoint {
        let lightness: Double
        let a: Double
        let b: Double

        var chroma: Double { hypot(a, b) }
    }

    private struct RenotationSample {
        let astmHue: Double
        let value: Double
        let chroma: Double
        let lab: LabPoint
    }

    private struct WeightedSample {
        let sample: RenotationSample
        let squaredDistance: Double
    }

    private static let illuminantCWhite = (x: 0.9807059717, y: 1.0, z: 1.1822494939)
    private static let neighborCount = 12
    private static let neutralChromaThreshold = 0.5
    private static let renotationSamples = loadRenotationSamples()
    private static let colorimetrySamples = loadColorimetrySamples()

    static func convert(reflectanceSpectrum: [SpectralSample]) -> MunsellNotation? {
        guard let xyz = illuminantCXYZ(reflectanceSpectrum: reflectanceSpectrum),
              !renotationSamples.isEmpty else {
            return nil
        }

        let normalizedXYZ = (
            x: xyz.first / 100,
            y: xyz.second / 100,
            z: xyz.third / 100
        )
        let target = labPoint(from: normalizedXYZ, white: illuminantCWhite)
        guard [normalizedXYZ.x, normalizedXYZ.y, normalizedXYZ.z,
               target.lightness, target.a, target.b]
            .allSatisfy(\.isFinite) else {
            return nil
        }

        let value = munsellValue(forLuminanceFactor: xyz.second)
        let nearbyByValue = renotationSamples.filter { abs($0.value - value) <= 1.1 }
        let pool = nearbyByValue.isEmpty ? renotationSamples : nearbyByValue
        let neighbors = pool
            .map { sample in
                let deltaL = sample.lab.lightness - target.lightness
                let deltaA = sample.lab.a - target.a
                let deltaB = sample.lab.b - target.b
                return WeightedSample(
                    sample: sample,
                    squaredDistance: deltaL * deltaL + deltaA * deltaA + deltaB * deltaB
                )
            }
            .sorted { $0.squaredDistance < $1.squaredDistance }
            .prefix(neighborCount)

        guard !neighbors.isEmpty else { return nil }

        var sine = 0.0
        var cosine = 0.0
        var chromaScale = 0.0
        var totalWeight = 0.0
        for neighbor in neighbors {
            let weight = 1 / (neighbor.squaredDistance + 0.000_001)
            let angle = neighbor.sample.astmHue * 2 * .pi / 100
            sine += weight * sin(angle)
            cosine += weight * cos(angle)
            chromaScale += weight * neighbor.sample.chroma / max(neighbor.sample.lab.chroma, 0.000_001)
            totalWeight += weight
        }

        guard totalWeight.isFinite, totalWeight > 0 else { return nil }
        let chroma = min(max(target.chroma * chromaScale / totalWeight, 0), 50)
        if chroma < neutralChromaThreshold {
            return MunsellNotation(hue: nil, hueDesignator: nil, value: value, chroma: 0)
        }

        var astmHue = atan2(sine, cosine) * 100 / (2 * .pi)
        if astmHue <= 0 {
            astmHue += 100
        }
        let hue = hueNotation(fromASTMHue: astmHue)
        return MunsellNotation(
            hue: hue.prefix,
            hueDesignator: hue.designator,
            value: value,
            chroma: chroma
        )
    }

    static func illuminantCXYZ(
        reflectanceSpectrum: [SpectralSample]
    ) -> Vector3? {
        let reflectancePoints = normalizedReflectancePoints(reflectanceSpectrum)
        guard reflectancePoints.count >= 2,
              let firstReflectance = reflectancePoints.first,
              let lastReflectance = reflectancePoints.last,
              let firstReference = colorimetrySamples.first,
              let lastReference = colorimetrySamples.last else {
            return nil
        }

        let lowerBound = max(firstReflectance.wavelength, firstReference.wavelength)
        let upperBound = min(lastReflectance.wavelength, lastReference.wavelength)
        guard upperBound - lowerBound >= 1 else { return nil }

        var wavelengths = [lowerBound]
        wavelengths.append(contentsOf: colorimetrySamples.lazy
            .map(\.wavelength)
            .filter { $0 > lowerBound && $0 < upperBound })
        wavelengths.append(upperBound)

        var xIntegral = 0.0
        var yIntegral = 0.0
        var zIntegral = 0.0
        var normalizationIntegral = 0.0
        var previous: (wavelength: Double, x: Double, y: Double,
                       z: Double, normalization: Double)?

        for wavelength in wavelengths {
            guard let reflectance = interpolateReflectance(
                reflectancePoints,
                at: wavelength
            ), let reference = interpolateColorimetry(at: wavelength) else {
                return nil
            }

            let scaledReflectance = max(reflectance / 100, 0)
            let current = (
                wavelength: wavelength,
                x: scaledReflectance * reference.illuminant * reference.xBar,
                y: scaledReflectance * reference.illuminant * reference.yBar,
                z: scaledReflectance * reference.illuminant * reference.zBar,
                normalization: reference.illuminant * reference.yBar
            )
            if let previous {
                let interval = current.wavelength - previous.wavelength
                xIntegral += interval * (previous.x + current.x) / 2
                yIntegral += interval * (previous.y + current.y) / 2
                zIntegral += interval * (previous.z + current.z) / 2
                normalizationIntegral += interval
                    * (previous.normalization + current.normalization) / 2
            }
            previous = current
        }

        guard normalizationIntegral.isFinite,
              normalizationIntegral > 0 else {
            return nil
        }
        let scale = 100 / normalizationIntegral
        let result = Vector3(
            first: xIntegral * scale,
            second: yIntegral * scale,
            third: zIntegral * scale
        )
        return [result.first, result.second, result.third].allSatisfy(\.isFinite)
            ? result
            : nil
    }

    private static func normalizedReflectancePoints(
        _ spectrum: [SpectralSample]
    ) -> [ReflectancePoint] {
        let sorted = spectrum
            .filter { $0.wavelength.isFinite && $0.value.isFinite }
            .sorted { $0.wavelength < $1.wavelength }
        var points: [ReflectancePoint] = []
        for sample in sorted {
            let point = ReflectancePoint(
                wavelength: sample.wavelength,
                value: sample.value
            )
            if let last = points.last,
               abs(last.wavelength - point.wavelength) < 0.000_000_001 {
                points[points.count - 1] = point
            } else {
                points.append(point)
            }
        }
        return points
    }

    private static func interpolateReflectance(
        _ points: [ReflectancePoint],
        at wavelength: Double
    ) -> Double? {
        guard let first = points.first,
              let last = points.last,
              wavelength >= first.wavelength,
              wavelength <= last.wavelength else {
            return nil
        }
        var lower = 0
        var upper = points.count - 1
        while lower + 1 < upper {
            let midpoint = (lower + upper) / 2
            if points[midpoint].wavelength <= wavelength {
                lower = midpoint
            } else {
                upper = midpoint
            }
        }
        let firstPoint = points[lower]
        if abs(firstPoint.wavelength - wavelength) < 0.000_000_001 {
            return firstPoint.value
        }
        let secondPoint = points[upper]
        let interval = secondPoint.wavelength - firstPoint.wavelength
        guard interval > 0 else { return firstPoint.value }
        let fraction = (wavelength - firstPoint.wavelength) / interval
        return firstPoint.value + fraction * (secondPoint.value - firstPoint.value)
    }

    private static func interpolateColorimetry(
        at wavelength: Double
    ) -> ColorimetrySample? {
        guard let first = colorimetrySamples.first,
              let last = colorimetrySamples.last,
              wavelength >= first.wavelength,
              wavelength <= last.wavelength else {
            return nil
        }
        let position = wavelength - first.wavelength
        let lowerIndex = min(max(Int(floor(position)), 0), colorimetrySamples.count - 1)
        let lower = colorimetrySamples[lowerIndex]
        if lowerIndex == colorimetrySamples.count - 1
            || abs(lower.wavelength - wavelength) < 0.000_000_001 {
            return lower
        }
        let upper = colorimetrySamples[lowerIndex + 1]
        let fraction = (wavelength - lower.wavelength)
            / (upper.wavelength - lower.wavelength)
        return ColorimetrySample(
            wavelength: wavelength,
            illuminant: lower.illuminant
                + fraction * (upper.illuminant - lower.illuminant),
            xBar: lower.xBar + fraction * (upper.xBar - lower.xBar),
            yBar: lower.yBar + fraction * (upper.yBar - lower.yBar),
            zBar: lower.zBar + fraction * (upper.zBar - lower.zBar)
        )
    }

    private static func hueNotation(fromASTMHue astmHue: Double) -> (prefix: Double, designator: String) {
        let designators = ["R", "YR", "Y", "GY", "G", "BG", "B", "PB", "P", "RP"]
        var normalized = astmHue.truncatingRemainder(dividingBy: 100)
        if normalized <= 0 { normalized += 100 }
        var segment = Int(floor(normalized / 10))
        var prefix = normalized - Double(segment) * 10
        if prefix < 0.000_001 {
            segment = (segment + designators.count - 1) % designators.count
            prefix = 10
        } else {
            segment %= designators.count
        }
        return (prefix, designators[segment])
    }

    private static func munsellValue(forLuminanceFactor luminance: Double) -> Double {
        let target = min(max(luminance, 0), 100)
        var lower = 0.0
        var upper = 10.0
        for _ in 0..<60 {
            let candidate = (lower + upper) / 2
            if astmLuminance(forMunsellValue: candidate) < target {
                lower = candidate
            } else {
                upper = candidate
            }
        }
        return (lower + upper) / 2
    }

    private static func astmLuminance(forMunsellValue value: Double) -> Double {
        0.00081939 * pow(value, 5)
            - 0.020484 * pow(value, 4)
            + 0.23352 * pow(value, 3)
            - 0.22533 * pow(value, 2)
            + 1.1914 * value
    }

    private static func labPoint(
        from xyz: (x: Double, y: Double, z: Double),
        white: (x: Double, y: Double, z: Double)
    ) -> LabPoint {
        let fx = labPivot(xyz.x / white.x)
        let fy = labPivot(xyz.y / white.y)
        let fz = labPivot(xyz.z / white.z)
        return LabPoint(
            lightness: 116 * fy - 16,
            a: 500 * (fx - fy),
            b: 200 * (fy - fz)
        )
    }

    private static func labPivot(_ value: Double) -> Double {
        let epsilon = 216.0 / 24_389.0
        let kappa = 24_389.0 / 27.0
        return value > epsilon ? cbrt(value) : (kappa * value + 16) / 116
    }

    private static func loadRenotationSamples() -> [RenotationSample] {
        guard let url = Bundle.module.url(
            forResource: "MunsellRenotationAll",
            withExtension: "csv"
        ), let content = try? String(contentsOf: url, encoding: .utf8) else {
            return []
        }

        return content.split(whereSeparator: \.isNewline).dropFirst().compactMap { line in
            let fields = line.split(separator: ",", omittingEmptySubsequences: false)
            guard fields.count == 6,
                  let value = Double(fields[1]),
                  let chroma = Double(fields[2]),
                  let x = Double(fields[3]),
                  let y = Double(fields[4]),
                  let rawY = Double(fields[5]),
                  abs(y) > 0.000_000_001,
                  let astmHue = astmHue(from: String(fields[0])) else {
                return nil
            }

            let luminance = rawY * 0.975 / 100
            let xyz = (
                x: x * luminance / y,
                y: luminance,
                z: (1 - x - y) * luminance / y
            )
            let sampleLab = labPoint(from: xyz, white: illuminantCWhite)
            guard [sampleLab.lightness, sampleLab.a, sampleLab.b].allSatisfy(\.isFinite),
                  sampleLab.chroma > 0.000_001 else {
                return nil
            }
            return RenotationSample(
                astmHue: astmHue,
                value: value,
                chroma: chroma,
                lab: sampleLab
            )
        }
    }

    private static func astmHue(from label: String) -> Double? {
        let designators = ["BG", "GY", "YR", "RP", "PB", "B", "G", "Y", "R", "P"]
        let codes = ["BG": 2, "GY": 4, "YR": 6, "RP": 8, "PB": 10,
                     "B": 1, "G": 3, "Y": 5, "R": 7, "P": 9]
        guard let designator = designators.first(where: { label.hasSuffix($0) }),
              let code = codes[designator],
              let prefix = Double(label.dropLast(designator.count)) else {
            return nil
        }
        let preciseHue = Double(10 * ((7 - code + 10) % 10)) + prefix
        return abs(preciseHue) < 0.000_001 ? 100 : preciseHue
    }

    private static func loadColorimetrySamples() -> [ColorimetrySample] {
        guard let url = Bundle.module.url(
            forResource: "MunsellColorimetryCIE1931",
            withExtension: "csv"
        ), let content = try? String(contentsOf: url, encoding: .utf8) else {
            return []
        }

        return content.split(whereSeparator: \.isNewline).dropFirst().compactMap { line in
            let fields = line.split(separator: ",", omittingEmptySubsequences: false)
            guard fields.count == 5,
                  let wavelength = Double(fields[0]),
                  let illuminant = Double(fields[1]),
                  let xBar = Double(fields[2]),
                  let yBar = Double(fields[3]),
                  let zBar = Double(fields[4]) else {
                return nil
            }
            return ColorimetrySample(
                wavelength: wavelength,
                illuminant: illuminant,
                xBar: xBar,
                yBar: yBar,
                zBar: zBar
            )
        }
        .sorted { $0.wavelength < $1.wavelength }
    }
}
